sdk/metrics/src/main/scala/org/typelevel/otel4s/sdk/metrics/data/ExemplarData.scala

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2024 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk.metrics.data
+
+import cats.Hash
+import cats.Show
+import cats.syntax.foldable._
+import org.typelevel.otel4s.Attributes
+import scodec.bits.ByteVector
+
+import scala.concurrent.duration.FiniteDuration
+
+/** An exemplar is a recorded value that associates
+  * [[ExemplarData.TraceContext]] (extracted from the
+  * [[org.typelevel.otel4s.sdk.context.Context Context]]) to a metric event.
+  *
+  * It allows linking trace details with metrics.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/metrics/data-model/#exemplars]]
+  */
+sealed trait ExemplarData {
+  type Value
+
+  /** A set of filtered attributes that provide additional insight about the
+    * measurement.
+    */
+  def filteredAttributes: Attributes
+
+  /** The time of the observation.
+    */
+  def timestamp: FiniteDuration
+
+  /** The trace associated with a recording.
+    */
+  def traceContext: Option[ExemplarData.TraceContext]
+
+  /** The recorded value.
+    */
+  def value: Value
+
+  override final def hashCode(): Int =
+    Hash[ExemplarData].hash(this)
+
+  override final def equals(obj: Any): Boolean =
+    obj match {
+      case other: ExemplarData => Hash[ExemplarData].eqv(this, other)
+      case _                   => false
+    }
+
+  override final def toString: String =
+    Show[ExemplarData].show(this)
+}
+
+object ExemplarData {
+
+  sealed trait LongExemplar extends ExemplarData { type Value = Long }
+
+  sealed trait DoubleExemplar extends ExemplarData { type Value = Double }
+
+  /** The trace information.
+    */
+  sealed trait TraceContext {
+    def traceId: ByteVector
+    def spanId: ByteVector
```

Comment:
`TraceContext` is a minimal version of `SpanContext`. That way, `sdk-metrics` does not need to depend on the `core-trace`.

Reply:
would it make sense for `TraceContext` to be in `core-common` and for `SpanContext` to extend it?

Reply:
Hard to say, to be honest. Once we have a functioning prototype, we can revisit this idea.

It sounds interesting and may solve some problems, actually.